New Db parameter 'UsePreparedStatement'

szampier opened the pull request:
@rtobar at ESO we are testing NGAS against a Sybase DB and we have found some issues with prepared statements. We have tried both the official Python Sybase driver `sybpydb` and `pyodbc` and both seem to have problems with prepared statements. If we don't manage to solve these issues I would propose to implement an option to combine SQL and code in the application and send "static" queries to the server.
This is just an initial implementation for your consideration.
Thanks, Stefano

rtobar commented on src/ngamsCore/ngamsLib/ngamsDbCore.py:
```diff
@@ -819,6 +822,10 @@ def _data_to_bind(self, data):
 
     def _prepare_query(self, sql, args):
 
+        if not self._use_prepared_statement:
+            markers = ["'{}'" if isinstance(arg, str) else "{}" for arg in args]
+            return sql.format(*markers).format(*args), ()
```
I think this logic is a tad too simple to work correctly. It will work for parameters that are strings themselves, but, for example, not when the argument is a list of strings (is that even a possibility? I'm actually not sure). And it needs to escape single quotes in the strings too.

It'd be great to actually test this too. A suggestion would be to add another job to the CI job matrix where we locally modify the `test/src/ngamsCfg.xml` file to add `UsePreparedStatement="true"`, then let the tests run as usual.

szampier replied:
@rtobar good points!
* logic was improved to handle `None` values and single quotes
* I added a test to the matrix for `Python3.9`, `MySQL` and `UsePreparedStatement="false"`. Now all tests are passing!
* in my last commit I fixed exiting code not passing arguments to `cursor.execute(sql)` if the arguments are empty. This logic was already implemented in `transaction` but not in `ngamsDbCursor`

rtobar reviewed (commented):
Thanks @szampier, this is looking good already! I had seen the first iteration yesterday, but forgot to come back to it, and I see you already took care of passing the option down to the DB object when created.

Could you add an item under `doc/changelog.rst` (under a new `.. rubric:: Development`) to mention this new option is being added? Similarly, editing the `cfg/sample_server_config.xml` file to mention this would be great.

In my inline comment I suggest we actually test this to verify it's working as expected. If we see the pipeline working I'll be happy to merge.

rtobar commented on src/ngamsCore/ngamsLib/ngamsConfig.py:
```diff
@@ -962,6 +962,17 @@ def getDbUseFileIgnore(self):
+        Indicates whether to use prepared statements (default) or call SQL directly
+        from the application source code in a way that combines code and data"
+        """
+        val = self.getVal("Db[1].UsePreparedStatement")
```
A nit-pick: could we name this `UsePreparedStatements` (plural rather than singular)? Feels more appropriate given how it applies to all queries, even the docstring above refers to them in plural.

szampier replied:
Done!

rtobar commented on .github/workflows/build-and-test-ngas.yml:
```diff
@@ -31,6 +31,10 @@ jobs:
           test_number: 5
+        - python: 3.9
+          db: mysql
+          sql_mode: no_prepared_statement
```
Suggestion: name this parameter `use_prepared_statements` (using plural here, but see my other comment), rename the env variable `SQL_MODE` to `USE_PREPARED_STATEMENTS` in the `Run NGAS tests` step, finally calculate `use_prepared_statements_attr='UsePreparedStatements="'${USE_PREPARED_STATEMENTS:-true}'"'` in the `run_tests.sh` script and always append the attribute to the `NGAS_TESTDB` definition. That makes it a bit less mysql-specific, and keeps the naming consistent across all the layers.

BTW, have you run locally the test suite with this new option set against a Sybase DB?

szampier replied:
I've implemented all your suggestions. The reason for having `sql_mode=no_prepared_statements` was to display something more meaningful than `false` in the test execution summary in github, but I agree that consistency across the code is probably more important.
I have not run the tests against Sybase but I'm running manual tests on a pre-production system and this is how I found the issue with passing empty args to `cursor.execute()`

rtobar commented on src/ngamsCore/ngamsLib/ngamsDbCore.py:
```diff
@@ -454,7 +454,10 @@ def __init__(self, pool, query, args):
             self.conn = pool.connection()
             self.cursor = self.conn.cursor()
-            self.cursor.execute(query, args)
+            if args:
```
nit-pick: could you copy over the comment that explains the need for this `if` from `transaction.execute`? Even better, if you could put this duplicated logic with a comment in a new non-public free-standing function that'd be great.

rtobar reviewed (commented):
This is looking great, thanks for those tests! I left some minor nit-picks and suggestions, but I'm happy if you brush them off. Let's add something to the changelog though, as suggested.

szampier commented:
@rtobar I've implemented your suggestions and added a reference to the new parameter to the change log and to the sample config file. We will keep testing against Sybase but so far it looks good. I've changed the PR status to Ready.

rtobar commented:
@szampier thanks for those further changes. Would you want me to hold on merging in case you find something else, or should I just go ahead? Your choice. The MR looks perfect to me :), so I'm happy to merge now if you say so.

rtobar commented:
Well, you actually marked the PR as Ready, so I'll be taking that as a "go". Merged into `master` now!
